challenges/day19_script.js

```diff
@@ -0,0 +1,122 @@
+const fs = require('fs');
+
+const fileName = '../inputs/day19_input.txt';
+
+class Rule {
+	constructor(id, values, patterns) {
+		this.id = id;
+		this.values = values;
+		this.patterns = patterns;
+	}
+}
+
+function run() {
+	const sum = parseInput();
+
+	console.log(sum);
+}
+
+function parseInput() {
+	const rl = fs.readFileSync(fileName).toString("utf-8").split("\r\n").map(l => l.trim());
+	const inputBreak = rl.indexOf(""); // where rules stop
+
+	let rules = new Map();
+
+	// rules
+	rl.slice(0, inputBreak).forEach(l => {
+		let rule = parseRule(l);
+		rules.set(rule.id, rule);
+	});
+
+	getValues(rules, 0);
+
+	// messages
+	let matches = 0;
+	rl.slice(inputBreak + 1).forEach(l => {
+		if (rules.get(0).values.some(v => v === l))
+			matches++;
+	});
+
+	return matches;
+}
+
+function parseRule(line) {
+	const idRegex = /^([0-9]+):/;
+	const patternRegex = /: ([0-9| ]+)$/;
+	const valueRegex = /: \"([a-z+])\"/;
+
+	const id = parseInt(idRegex.exec(line)[1]);
+	const patternExists = patternRegex.exec(line);
+	const pattern = patternExists !== null ? parsePatterns(patternExists[1]) : undefined;
+	const valueExists = valueRegex.exec(line);
+	const value = valueExists !== null ? valueExists[1] : undefined;
+
+	return new Rule(id, value ? [value] : [], pattern);
+}
+
+function parsePatterns(string) {
+	const patterns = new Set();
+
+	string.split("|").filter(s => s !== "").forEach(s => {
+		const numStrings = s.split(" ").filter(s => s !== "");
+		patterns.add(numStrings.map(n => parseInt(n)));
+	});
+
+	return patterns;
+}
+
+function getValues(rules, ruleId) {
+	const rule = rules.get(ruleId);
+
+	if (rule.values.length !== 0)
+		return rule.values;
+
+	let values = [];
+	rule.patterns.forEach(p => {
+		values.push([]);
+
+		for (var j = 0; j < p.length; j++) {
+			values[values.length - 1].push(getValues(rules, p[j]));
+		}
+	});
+
+	rule.values = parseValues(ruleId, values);
+
+	return rule.values;
+}
+
+function parseValues(ruleId, values) {
+	let combinedValues = [];
+
+	for (var i = 0; i < values.length; i++) {
+		const valI = values[i];
+
+		for (var j = 0; j < valI[0].length; j++) {
+			const jthStart = valI[0][j];
+
+			if (valI.length >= 2) {
+				for (var k = 0; k < valI[1].length; k++) {
+					const kthMiddle = valI[1][k];
+
+					if (valI.length >= 3) {
+						for (var l = 0; l < valI[2].length; l++) {
+							const lthEnd = valI[2][l];
+
+							combinedValues.push(jthStart + kthMiddle + lthEnd);
+						}
+					}
+					else
+						combinedValues.push(jthStart + kthMiddle);
+				}
+			}
+			else
+				combinedValues.push(jthStart);
+		}
+	}
+
+	console.log(ruleId, combinedValues);
+	return combinedValues;
+}
```

Comment:
This is bad. Enumerated all possible matches for each regex. Already inefficient, but part 2 cemented the inefficiency